core/drivers/stm32_cpu_opp.c

```diff
@@ -0,0 +1,509 @@
+				     PWR_CR1_MPU_RAM_LOW_SPEED);
+#endif
+
+		if (set_opp_voltage(cpu_opp.current_opp))
+			panic();
+
+		return res;
+	}
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result stm32_cpu_opp_set_level(unsigned int level)
+{
+	unsigned int current_level = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned int opp = 0;
+
+	assert(cpu_opp.dvfs);
+
+	DMSG("Set OPP to %ukHz", level);
+
+	if (!cpu_opp.dvfs)
+		return TEE_ERROR_GENERIC;
+
+	mutex_lock(&cpu_opp_mu);
+
+	current_level = stm32_cpu_opp_level(cpu_opp.current_opp);
+
+	if (level == current_level) {
+		mutex_unlock(&cpu_opp_mu);
+		return TEE_SUCCESS;
+	}
+
+	for (opp = 0; opp < cpu_opp.opp_count; opp++)
+		if (level == cpu_opp.dvfs[opp].freq_khz)
+			break;
+
+	if (opp == cpu_opp.opp_count) {
+		mutex_unlock(&cpu_opp_mu);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (level < current_level)
+		res = set_clock_then_voltage(opp);
+	else
+		res = set_voltage_then_clock(opp);
+
+	if (res)
+		EMSG("Failed to set OPP to %ukHz", level);
+	else
+		cpu_opp.current_opp = opp;
+
+	mutex_unlock(&cpu_opp_mu);
+
+	return res;
+}
+
+TEE_Result stm32_cpu_opp_read_level(unsigned int *level)
+{
+	if (!cpu_opp.opp_count) {
+		EMSG("No CPU OPP defined");
+		return TEE_ERROR_GENERIC;
+	}
+
+	*level = stm32_cpu_opp_level(cpu_opp.current_opp);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_cpu_opp_is_supported(const void *fdt, int subnode)
+{
+	const fdt32_t *cuint32 = NULL;
+	uint32_t opp = 0;
+
+	cuint32 = fdt_getprop(fdt, subnode, "opp-supported-hw", NULL);
+	if (!cuint32) {
+		DMSG("Can't find property opp-supported-hw");
+		return TEE_ERROR_GENERIC;
+	}
+
+	opp = fdt32_to_cpu(*cuint32);
+	if (!stm32mp_supports_cpu_opp(opp)) {
+		DMSG("Not supported opp-supported-hw %#"PRIx32, opp);
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result cpu_opp_pm(enum pm_op op, unsigned int pm_hint,
+			     const struct pm_callback_handle *hdl __unused)
+{
+	assert(op == PM_OP_SUSPEND || op == PM_OP_RESUME);
+
+	/* nothing to do if RCC clock tree is not lost */
+	if (!PM_HINT_IS_STATE(pm_hint, CONTEXT))
+		return TEE_SUCCESS;
+
+	if (op == PM_OP_RESUME) {
+		unsigned long clk_cpu = 0;
+		unsigned int opp = cpu_opp.current_opp;
+
+		DMSG("Resume to OPP %u", opp);
+
+		clk_cpu = clk_get_rate(cpu_opp.clock);
+		assert(clk_cpu);
+		if (cpu_opp.dvfs[opp].freq_khz * 1000 >= clk_cpu)
+			return set_voltage_then_clock(opp);
+		else
+			return set_clock_then_voltage(opp);
+	}
+
+	return TEE_SUCCESS;
+}
+DECLARE_KEEP_PAGER(cpu_opp_pm);
+
+static TEE_Result stm32_cpu_opp_get_dt_subnode(const void *fdt, int node)
+{
+	const fdt64_t *cuint64 = NULL;
+	const fdt32_t *cuint32 = NULL;
+	uint64_t freq_hz = 0;
+	uint64_t freq_khz = 0;
+	uint64_t freq_khz_opp_def = 0;
+	uint32_t volt_uv = 0;
+	unsigned long clk_cpu = 0;
+	unsigned int i = 0;
+	int subnode = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	cpu_opp.dvfs = calloc(CFG_STM32MP_OPP_COUNT, sizeof(*cpu_opp.dvfs));
```

Comment:
Since we always allocate `CFG_STM32MP_OPP_COUNT` cells, maybe worth to make the array size statically defined in the struct:
```patch
 struct cpu_opp {
 	unsigned int current_opp;
 	unsigned int opp_count;
 	struct clk *clock;
 	struct regulator *regul;
-	struct cpu_dvfs *dvfs;
+	struct cpu_dvfs dvfs[CFG_STM32MP_OPP_COUNT];
 	unsigned int sustained_freq_khz;
 };
```